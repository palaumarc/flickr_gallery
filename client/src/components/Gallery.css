.gallery {
    display: grid;
    grid-template-columns: repeat(auto-fill, minmax(250px, 1fr));
    grid-gap: 5px;
}

.gallery-photo {
    width: 100%;
    height: 250px;
    object-fit: cover;
}