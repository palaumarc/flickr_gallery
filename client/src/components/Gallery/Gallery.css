.gallery {
    display: grid;
    grid-template-columns: repeat(auto-fill, minmax(250px, 1fr));
    grid-gap: 5px;
}

.gallery-photo-container {
    position: relative;
    cursor: pointer;
}

.gallery-photo {
    width: 100%;
    height: 250px;
    object-fit: cover;
}

.overlay {
    position: absolute;
    top: 0;
    bottom: 0;
    left: 0;
    right: 0;
    height: 100%;
    width: 100%;
    opacity: 0;
    transition: .5s ease;
    background-color: rgb(173, 173, 173, 0.7);
  }
  
  .overlay:hover {
    opacity: 1;
  }
  
  .text {
    color: white;
    font-size: 20px;
    position: absolute;
    top: 50%;
    left: 50%;
    transform: translate(-50%, -50%);
    -ms-transform: translate(-50%, -50%);
    text-overflow: ellipsis;
    white-space: nowrap;
    overflow: hidden;
    text-align: center;
    width: 80%;
  }