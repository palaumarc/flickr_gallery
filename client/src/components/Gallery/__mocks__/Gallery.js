import React from "react";

const Gallery = props => <gallery-mock {...props}>{props.children}</gallery-mock>;
const GalleryPhoto = props => <gallery-photo {...props}>{props.children}</gallery-photo>;

Gallery.Photo = GalleryPhoto;

export default Gallery;
