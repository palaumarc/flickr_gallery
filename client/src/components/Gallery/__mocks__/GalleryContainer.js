import React from "react";

const GalleryContainer = props => <gallery-container-mock {...props}>{props.children}</gallery-container-mock>;

export default GalleryContainer;