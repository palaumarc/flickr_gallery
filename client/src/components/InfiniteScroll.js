import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';

import Spin from './Spin';

class InfiniteScroll extends Component {

  static propTypes = {
      loadMore: PropTypes.func.isRequired,
      hasMore: PropTypes.bool
  }

  static defaultProps = {
      hasMore: true
  }

  state = {
      isLoading: false
  }

  onScroll = () => {
      const { isLoading } = this.state;

      if (isLoading) return;

      // Checks that the page has scrolled to the bottom
      if (window.innerHeight + document.documentElement.scrollTop === document.documentElement.offsetHeight) {
        this.execLoadMore();
      }
  };

  execLoadMore = async () => {
      this.setState(prevState => ({...prevState, isLoading: true}));
      await this.props.loadMore()
      this.setState(prevState => ({...prevState, isLoading: false}));

      if (!this.props.hasMore) {
          document.removeEventListener('scroll', this.onScroll);
      }
  }

  async componentDidMount() {
    document.addEventListener('scroll', this.onScroll);

    // Keep loading until available height is filled or there are no more elements
    while (document.documentElement.offsetHeight < window.innerHeight && this.props.hasMore) {
      await this.execLoadMore();
    }
  }

  render() {
    return (
        <Fragment>
            {this.props.children}
            {this.state.isLoading ? <Spin /> : null}
        </Fragment>
    )
  }
}

export default InfiniteScroll;