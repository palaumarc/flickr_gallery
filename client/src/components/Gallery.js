import React from 'react';
import './Gallery.css';

const Gallery = ({ children }) => (
    <div className="gallery">
        {children}
    </div>
)

const GalleryPhoto = ({ photo, onClick }) => (
    <img className="gallery-photo" src={photo.previewUrl} onClick={onClick} alt='' />
)

Gallery.Photo = GalleryPhoto;

export default Gallery;