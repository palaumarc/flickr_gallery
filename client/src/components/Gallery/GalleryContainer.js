import React, { Component, Fragment } from 'react';
import InfiniteScroll from '../utils/InfiniteScroll';

import Gallery from './Gallery';
import PhotoDetail from './PhotoDetail';
import Lightbox from '../utils/Lightbox';
import Carousel from '../utils/Carousel';
import { fetchPhotos } from '../../services/ApiCalls';

const PHOTOS_PER_PAGE = 15;

class GalleryContainer extends Component {

    state = {
        photos: [],
        hasMorePhotos: true,
        currentPage: 0,
        selectedPhotoIndex: null
    }

    onClickPhoto = selectedPhotoIndex => this.setState(prevState => ({...prevState, selectedPhotoIndex}));
    
    onClosePhotoDetail = () => this.setState(prevState => ({...prevState, selectedPhotoIndex: null}));

    loadPhotos = async () => {
        const pageToLoad = this.state.currentPage + 1;
        const {photos, hasMore} = await fetchPhotos(pageToLoad, PHOTOS_PER_PAGE);
        this.setState(prevState => ({
            ...this.prevState, 
            currentPage: pageToLoad, 
            photos: prevState.photos.concat(photos),
            hasMorePhotos: hasMore
        }));
    }

    render() {

        const { photos, selectedPhotoIndex, hasMorePhotos } = this.state;

        return (
            <Fragment>
                <InfiniteScroll loadMore={this.loadPhotos} hasMore={hasMorePhotos}>
                    <Gallery>
                        {photos.map((photo, index) => <Gallery.Photo key={`${photo.id}-${index}`} photo={photo} onClick={() => this.onClickPhoto(index)}/>)}
                    </Gallery>
                </InfiniteScroll>
                {selectedPhotoIndex !== null ?
                <Lightbox onClickCloseButton={this.onClosePhotoDetail}>
                    <Carousel startIndex={selectedPhotoIndex}>
                        {photos.map((photo, index) => <PhotoDetail key={`${photo.id}-${index}`} photo={photo} />)}
                    </Carousel>
                </Lightbox> 
                : null}
            </Fragment>
        );
    }
}

export default GalleryContainer;
