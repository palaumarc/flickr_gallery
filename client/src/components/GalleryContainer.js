import React, { Component, Fragment } from 'react';
import InfiniteScroll from './InfiniteScroll';

import Gallery from './Gallery';
import PhotoDetail from './PhotoDetail';
import { fetchPhotos } from '../services/ApiCalls';

const PHOTOS_PER_PAGE = 15;

class GalleryContainer extends Component {

    state = {
        photos: [],
        currentPage: 0,
        selectedPhotoIndex: null
    }

    onClickPhoto = selectedPhotoIndex => this.setState({...this.state, selectedPhotoIndex});
    
    onClosePhotoDetail = () => this.setState({...this.state, selectedPhotoIndex: null});

    selectNextPhoto = () => {
        const newIndex = this.state.selectedPhotoIndex + 1;
        this.setState({...this.state, selectedPhotoIndex: newIndex});
    }

    selectPreviousPhoto = () => {
        const newIndex = this.state.selectedPhotoIndex - 1;
        this.setState({...this.state, selectedPhotoIndex: newIndex});
    }

    loadFunc = async () => {
        const pageToLoad = this.state.currentPage + 1;
        const newPhotos = await fetchPhotos(pageToLoad, PHOTOS_PER_PAGE);
        this.setState(prevState => ({
            ...this.prevState, 
            currentPage: pageToLoad, 
            photos: prevState.photos.concat(newPhotos)
        }));
    }

    render() {

        const { photos, selectedPhotoIndex } = this.state;

        return (
            <Fragment>
                <InfiniteScroll
                    loadMore={this.loadFunc}
                    hasMore={true}
                >
                    <Gallery>
                        {photos.map((photo, index) => <Gallery.Photo key={`${photo.id}-${index}`} photo={photo} onClick={() => this.onClickPhoto(index)}/>)}
                    </Gallery>
                </InfiniteScroll>
                {selectedPhotoIndex !== null ? 
                <PhotoDetail 
                    photo={photos[selectedPhotoIndex]} 
                    onClose={this.onClosePhotoDetail}
                    previousControl={selectedPhotoIndex > 0 ? <PhotoDetail.PreviousPhotoArrow onClick={this.selectPreviousPhoto}/> : null}
                    nextControl={selectedPhotoIndex < photos.length - 1 ? <PhotoDetail.NextPhotoArrow onClick={this.selectNextPhoto}/> : null}
                /> 
                : null}
            </Fragment>
        );
    }
}

export default GalleryContainer;
