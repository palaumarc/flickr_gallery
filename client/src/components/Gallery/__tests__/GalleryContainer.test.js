import React from "react";
import renderer from "react-test-renderer";
import ReactTestUtils from "react-dom/test-utils";

import GalleryContainer from "../GalleryContainer";
import Gallery from "../Gallery";
import InfiniteScroll from "../../utils/InfiniteScroll";
import { fetchPhotos } from '../../../services/ApiCalls';
import Lightbox from "../../utils/Lightbox";

jest.mock('../Gallery');
jest.mock('../PhotoDetail');
jest.mock('../../utils/Carousel');
jest.mock('../../utils/InfiniteScroll');
jest.mock('../../utils/Lightbox');
jest.mock('../../../services/ApiCalls');

const photosMock = [
    { id: 1 },
    { id: 2}
];

describe('Component is rendered.', () => {

    beforeEach(() => {
        jest.clearAllMocks();
    })

    test('Snapshot', async () => {
        const tree = renderer.create(<GalleryContainer />);
        expect(tree.toJSON()).toMatchSnapshot();
    })

    describe('InfiniteScroll triggers loadMore().', () => {

        test('Snapshot', async () => {
            fetchPhotos.mockReturnValue(Promise.resolve(photosMock));
            const tree = renderer.create(<GalleryContainer />);
            await tree.root.findByType(InfiniteScroll).props.loadMore();
            expect(tree.toJSON()).toMatchSnapshot();
        })

        test('FetchPhotos should have been called one time with expected parameters', async () => {
            fetchPhotos.mockReturnValue(Promise.resolve(photosMock));
            const tree = renderer.create(<GalleryContainer />);
            await tree.root.findByType(InfiniteScroll).props.loadMore();

            const expectedFetchPage = 1;
            const expectedFetchPerPage = 15;

            expect(fetchPhotos).toHaveBeenCalledTimes(1);
            expect(fetchPhotos).toHaveBeenCalledWith(expectedFetchPage, expectedFetchPerPage);
        })

        test('InfiniteScroll trigger loadMore() for second time. FetchPhotos should have been called two times with expected parameters', async () => {
            fetchPhotos.mockReturnValue(Promise.resolve(photosMock));
            const tree = renderer.create(<GalleryContainer />);
            await tree.root.findByType(InfiniteScroll).props.loadMore();
            await tree.root.findByType(InfiniteScroll).props.loadMore();

            const expectedFetchPage = 2;
            const expectedFetchPerPage = 15;

            expect(fetchPhotos).toHaveBeenCalledTimes(2);
            expect(fetchPhotos).toHaveBeenLastCalledWith(expectedFetchPage, expectedFetchPerPage);
        })

        describe('Gallery first photo is clicked.', () => {

            test('Snapshot', async () => {
                fetchPhotos.mockReturnValue(Promise.resolve(photosMock));
                const tree = renderer.create(<GalleryContainer />);
                await tree.root.findByType(InfiniteScroll).props.loadMore();
                tree.root.findAllByType(Gallery.Photo)[0].props.onClick();
                expect(tree.toJSON()).toMatchSnapshot();
            })

            describe('Lightbox is close button is clicked.', () => {

                test('Snapshot', async () => {
                    fetchPhotos.mockReturnValue(Promise.resolve(photosMock));
                    const tree = renderer.create(<GalleryContainer />);
                    await tree.root.findByType(InfiniteScroll).props.loadMore();
                    tree.root.findAllByType(Gallery.Photo)[0].props.onClick();
                    tree.root.findByType(Lightbox).props.onClickCloseButton();
                    expect(tree.toJSON()).toMatchSnapshot();
                })

            })

        })

    })

})