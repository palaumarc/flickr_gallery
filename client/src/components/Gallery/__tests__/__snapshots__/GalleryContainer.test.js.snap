// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component is rendered. InfiniteScroll triggers loadMore(). Gallery first photo is clicked. Lightbox is close button is clicked. Snapshot 1`] = `
<infinite-scroll-mock
  hasMore={true}
  loadMore={[Function]}
>
  <gallery-mock>
    <gallery-photo
      onClick={[Function]}
      photo={
        Object {
          "id": 1,
        }
      }
    />
    <gallery-photo
      onClick={[Function]}
      photo={
        Object {
          "id": 2,
        }
      }
    />
  </gallery-mock>
</infinite-scroll-mock>
`;

exports[`Component is rendered. InfiniteScroll triggers loadMore(). Gallery first photo is clicked. Snapshot 1`] = `
Array [
  <infinite-scroll-mock
    hasMore={true}
    loadMore={[Function]}
>
    <gallery-mock>
        <gallery-photo
            onClick={[Function]}
            photo={
                Object {
                    "id": 1,
                  }
            }
        />
        <gallery-photo
            onClick={[Function]}
            photo={
                Object {
                    "id": 2,
                  }
            }
        />
    </gallery-mock>
</infinite-scroll-mock>,
  <lightbox-mock
    onClickCloseButton={[Function]}
>
    <carousel-mock
        startIndex={0}
    >
        <photo-detail-mock
            photo={
                Object {
                    "id": 1,
                  }
            }
        />
        <photo-detail-mock
            photo={
                Object {
                    "id": 2,
                  }
            }
        />
    </carousel-mock>
</lightbox-mock>,
]
`;

exports[`Component is rendered. InfiniteScroll triggers loadMore(). Snapshot 1`] = `
<infinite-scroll-mock
  hasMore={true}
  loadMore={[Function]}
>
  <gallery-mock>
    <gallery-photo
      onClick={[Function]}
      photo={
        Object {
          "id": 1,
        }
      }
    />
    <gallery-photo
      onClick={[Function]}
      photo={
        Object {
          "id": 2,
        }
      }
    />
  </gallery-mock>
</infinite-scroll-mock>
`;

exports[`Component is rendered. Snapshot 1`] = `
<infinite-scroll-mock
  hasMore={true}
  loadMore={[Function]}
>
  <gallery-mock />
</infinite-scroll-mock>
`;
